### PR TITLE
fix: process delayed assignment metavariables correctly in `Lean.Meta.Closure`

### LINKED ISSUE #5925: A "declaration has metavariables" error
### Prerequisites
Please put an X between the brackets as you perform the following steps:

* [x] Check that your issue is not already filed.
* [x] Reduce the issue to a minimal, self-contained, reproducible test case.
* [x] Test your test case against the latest nightly release.

### Description
The following code

```lean
structure ValVar (D: Type) where
  d: D
  x: Nat

def Set (T : Type) := T → Prop

structure Salg where
  D: Type
  I: Nat

def natSalg: Salg := { D := Nat, I := 42 }

inductive HasMem (salg: Salg): Set (Set (ValVar salg.D))
| hasMem
  (set: Set (ValVar salg.D))
  (isElem: set ⟨d, x⟩)
:
  HasMem salg set

def valVarSet: Set (ValVar Nat) :=
  fun ⟨d, x⟩ => x = 0 ∧ d = 0 ∧ d ∉ []

def valVarSetHasMem: HasMem natSalg valVarSet :=
  HasMem.hasMem
    valVarSet
    (show valVarSet _ from ⟨rfl, rfl, nofun⟩)
```

throws `(kernel) declaration has metavariables 'valVarSetHasMem'`.

There is a list of things that make the issue go away (which is also why I couldn't come up with a more specific title -- I've no idea what the underlying cause is :sorry:):

* removing `x` from ValVar, along with its occurences in the rest of the program
* removing `I` from `Salg`
* removing the `salg` param from `HasMem` & replacing occurences with `natSalg`
* removing `d ∉ []` from the definition of `valVarSet`
* replacing `HasMem.hasMem valVarSet` with `HasMem.hasMem _` in `valVarSetHasMem`.
* replacing `show valVarSet _ from` with `show _ from`, `show valVarSet ⟨0, 0⟩ from`, or removing it entirely.

**Expected behavior:** The code throws no errors.

### Versions
Version: Lean 4.14.0-nightly-2024-11-01
Platform: live.lean-lang.org

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

### LINKED ISSUE #6354: Unexpected kernel error involving two structures
### Prerequisites

Please put an X between the brackets as you perform the following steps:

* [x] Check that your issue is not already filed:
      https://github.com/leanprover/lean4/issues
* [x] Reduce the issue to a minimal, self-contained, reproducible test case.
      Avoid dependencies to Mathlib or Batteries.
* [x] Test your test case against the latest nightly release, for example on
      https://live.lean-lang.org/#project=lean-nightly
      (You can also use the settings there to switch to “Lean nightly”)

### Description
```lean
structure A where
  p: Prop
  q: True

structure B extends A where
  q': p → True

example: B where
  p := True ∧ True
  q := by exact True.intro
  q' := λ ⟨_,_⟩ ↦ True.intro
```

**Expected behavior:** No errors.

**Actual behavior:** `(kernel) declaration has metavariables '_example'`

Mysteriously (to me), any one of the following changes make the error disappear:
- replacing `by exact True.intro` by `True.intro`
- replacing `⟨_,_⟩` by `_`
- moving any of `p`, `q` into `B` (or `q'` into `A`)
 
I originally encountered this example when playing with Mathlib definitions and then simplified it. I don't see how it could be reduced further.

### Context

I asked about this on [Zulip](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/Mysterious.20error.20message).

A similar issue is [#5925](https://github.com/leanprover/lean4/pull/5925) (pointed out by @kim-em).

If these two issues indeed have the same underlying cause, the present example is maybe simpler to investigate.

### Versions

Originally tested on Lean version 4.15.0-nightly-2024-11-15, also still persists on today's version at https://live.lean-lang.org/:
```
Lean 4.16.0-nightly-2024-12-10
Target: x86_64-unknown-linux-gnu
```

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

### PR DESCRIPTION
**Pull request by kmill**
This PR fixes a bug in `Lean.Meta.Closure` that would introduce under-applied delayed assignment metavariables, which would keep them from ever getting instantiated. This bug affected `match` elaboration when the expected type contained postponed elaboration problems, for example tactic blocks.

Closes #5925, closes #6354

### CODE REVIEW

File: src/Lean/Meta/Closure.lean [hunk 1]
```diff
@@ -203,9 +203,21 @@ partial def collectExprAux (e : Expr) : ClosureM Expr := do
...
+    /-
+    If the metavariable is a delayed assignment, it would be an error to push it onto `exprMVarArgs` directly,
+    since the assumption is that delayed assignment metavariables are always fully applied (for example `instantiateMVars` assumes this).
+    The need to handle delayed assignment metavariables was uncovered in issue https://github.com/leanprover/lean4/issues/6354
+    -/
+    let e' ←
+      if let some { fvars, .. } ← getDelayedMVarAssignment? mvarId then
+        -- Eta expand `e` for the requisite number of arguments.
+        forallBoundedTelescope mvarDecl.type fvars.size fun args _ => do
+          mkLambdaFVars args <| mkAppN e args
```
Comment (leodemoura) [inline src/Lean/Meta/Closure.lean:215]: Have you considered other solutions that do not require eta-expansion for this issue, even if they apply only to special cases? If they are not worth pursuing, we should document that here.

Comment (leodemoura) [reply src/Lean/Meta/Closure.lean:215]: > It did cross my mind that I could try matching on the case where the delayed assignment metavariable is applied to fvars, but it wasn't clear to me that the added code complexity would be worth it. We would still need the eta expansion code path for correctness.

Let's add an extra comment documenting this decision.

Comment (kmill) [reply src/Lean/Meta/Closure.lean:228]: Added.

(I also tried to implement this suggestion, but I couldn't get it to work within the time I gave toward attempting it. Here's [the branch](https://github.com/leanprover/lean4/compare/master...kmill:fix_6354_fvars) with non-working code.)

Comment (leodemoura) [reply src/Lean/Meta/Closure.lean:228]: Thanks!

**Pull request merged**